From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ilia Alshanetsky <ilia@ilia.ws>
Date: Thu, 16 Jul 2026 13:15:28 +0000
Subject: [PATCH] Fix GHSA-ch8v-r6jh-4vvr: encode 0xFF in
 FILTER_SANITIZE_ENCODED

Backport of https://github.com/php/php-src/commit/56149186d81469fd7fb2f8444359ee81dd4a70f4
7.4 reference: https://github.com/shivammathur/php-src-backports/commit/f517508eb09d9e25c2c1862ffe69b6e91588d6f2

5.6: test uses SKIPIF instead of --EXTENSIONS--.
---
 ext/filter/sanitizing_filters.c                    |  2 +-
 ext/filter/tests/filter_sanitize_encoded_0xff.phpt | 12 ++++++++++++
 2 files changed, 13 insertions(+), 1 deletion(-)
 create mode 100644 ext/filter/tests/filter_sanitize_encoded_0xff.phpt

diff --git a/ext/filter/sanitizing_filters.c b/ext/filter/sanitizing_filters.c
index 9f3053e6..581667c8 100644
--- a/ext/filter/sanitizing_filters.c
+++ b/ext/filter/sanitizing_filters.c
@@ -71,7 +71,7 @@ static void php_filter_encode_url(zval *value, const unsigned char* chars, const
 	unsigned char *s = (unsigned char *)chars;
 	unsigned char *e = s + char_len;
 
-	memset(tmp, 1, sizeof(tmp)-1);
+	memset(tmp, 1, sizeof(tmp));
 
 	while (s < e) {
 		tmp[*s++] = 0;
diff --git a/ext/filter/tests/filter_sanitize_encoded_0xff.phpt b/ext/filter/tests/filter_sanitize_encoded_0xff.phpt
new file mode 100644
index 00000000..29cbda81
--- /dev/null
+++ b/ext/filter/tests/filter_sanitize_encoded_0xff.phpt
@@ -0,0 +1,12 @@
+--TEST--
+FILTER_SANITIZE_ENCODED percent-encodes 0xFF
+--SKIPIF--
+<?php if (!extension_loaded("filter")) die("skip"); ?>
+--FILE--
+<?php
+var_dump(filter_var("\xFF", FILTER_SANITIZE_ENCODED));
+var_dump(filter_var("\xFE\xFF\x00A", FILTER_SANITIZE_ENCODED));
+?>
+--EXPECT--
+string(3) "%FF"
+string(10) "%FE%FF%00A"
